pkg/ingester/retention.go

```diff
@@ -39,183 +66,351 @@ func defaultRetentionPolicy() retentionPolicy {
-	// The call is thread-safe: tenant can't be added or removed
-	// during the execution.
-	evictBlock(tenant string, b ulid.ULID, fn func() error) error
-}
+	logger        log.Logger
+	config        phlaredb.Config
+	policy        retentionPolicy
+	blockManager  fsBlockManager
+	volumeChecker diskutil.VolumeChecker
 
-func newRetentionPolicyEnforcer(logger log.Logger, blockEvicter blockEvicter, retentionPolicy retentionPolicy, dbConfig phlaredb.Config) *retentionPolicyEnforcer {
-	e := retentionPolicyEnforcer{
-		logger:          logger,
-		blockEvicter:    blockEvicter,
-		retentionPolicy: retentionPolicy,
-		dbConfig:        dbConfig,
-		stopCh:          make(chan struct{}),
-		fileSystem:      new(realFileSystem),
-		volumeChecker:   diskutil.NewVolumeChecker(retentionPolicy.MinFreeDisk*1024*1024*1024, retentionPolicy.MinDiskAvailablePercentage),
-	}
-	e.Service = services.NewBasicService(nil, e.running, e.stopping)
-	return &e
+	stop chan struct{}
+	wg   sync.WaitGroup
 }
 
-func (e *retentionPolicyEnforcer) running(ctx context.Context) error {
-	e.wg.Add(1)
-	retentionPolicyEnforcerTicker := time.NewTicker(e.retentionPolicy.EnforcementInterval)
+func (dc *diskCleaner) running(ctx context.Context) error {
+	dc.wg.Add(1)
+	ticker := time.NewTicker(dc.policy.EnforcementInterval)
 	defer func() {
-		retentionPolicyEnforcerTicker.Stop()
-		e.wg.Done()
+		ticker.Stop()
+		dc.wg.Done()
 	}()
+
+	var deleted int
+	var bytesDeleted int
+	var hasHighDiskUtilization bool
 	for {
-		// Enforce retention policy immediately at start.
-		level.Debug(e.logger).Log("msg", "enforcing retention policy")
-		if err := e.cleanupBlocksWhenHighDiskUtilization(ctx); err != nil {
-			level.Error(e.logger).Log("msg", "failed to enforce retention policy", "err", err)
+		deleted = dc.DeleteUploadedBlocks(ctx)
+		level.Debug(dc.logger).Log("msg", "cleaned uploaded blocks", "count", deleted)
+
+		deleted, bytesDeleted, hasHighDiskUtilization = dc.EnforceHighDiskUtilization(ctx)
+		if hasHighDiskUtilization {
+			level.Debug(dc.logger).Log(
+				"msg", "cleaned files after high disk utilization",
+				"deleted_blocks", deleted,
+				"deleted_bytes", bytesDeleted,
+			)
 		}
+
 		select {
-		case <-retentionPolicyEnforcerTicker.C:
+		case <-ticker.C:
 		case <-ctx.Done():
 			return nil
-		case <-e.stopCh:
+		case <-dc.stop:
 			return nil
 		}
 	}
 }
 
-func (e *retentionPolicyEnforcer) stopping(_ error) error {
-	close(e.stopCh)
-	e.wg.Wait()
+func (dc *diskCleaner) stopping(_ error) error {
+	close(dc.stop)
+	dc.wg.Wait()
 	return nil
 }
 
-func (e *retentionPolicyEnforcer) localBlocks(dir string) ([]*tenantBlock, error) {
-	blocks := make([]*tenantBlock, 0, 32)
-	tenants, err := fs.ReadDir(e.fileSystem, dir)
+// DeleteUploadedBlocks scans and deletes blocks on all tenants that have
+// already been deleted. It returns the number of blocks deleted.
+func (dc *diskCleaner) DeleteUploadedBlocks(ctx context.Context) int {
```

Comment:
We should make sure that the operation handles context cancellation and does not block the server shutdown. Potentially, clean up can take quite a while (especially on the first run), therefore I guess we could check context each time we deleted a block. Same for `EnforceHighDiskUtilization`

Reply:
In https://github.com/grafana/pyroscope/pull/2816/commits/0399b5920d5f06f59bf2bf09c2faa1df06baed9c, I added checks for context cancellations before reading or deleting blocks.